File: Helper/Core/Drag/AddMode/ModifiedDragOutputAddMode.m
//
// --------------------------------------------------------------------------
// ModifiedDragOutputAddMode.m
// Created for Mac Mouse Fix (https://github.com/noah-nuebling/mac-mouse-fix)
// Created by Noah Nuebling in 2022
// Licensed under the MMF License (https://github.com/noah-nuebling/mac-mouse-fix/blob/master/License)
// --------------------------------------------------------------------------
//

#import "ModifiedDragOutputAddMode.h"
#import "Remap.h"

@implementation ModifiedDragOutputAddMode

/// Vars

static ModifiedDragState *_drag;
static NSDictionary *_addModePayload; /// Payload to send to the mainApp. Only used with modified drag of type kMFModifiedDragTypeAddModeFeedback.
static BOOL _didConclude;

/// Interface

+ (void)initializeWithDragState:(ModifiedDragState *)dragStateRef {
    
    _drag = dragStateRef;
    
    /// Prepare payload to send to mainApp during AddMode. See Remap -> AddMode for context
    NSMutableDictionary *payload = _drag->effectDict.mutableCopy; /// Probably already mutable. See RemapSwizzler.
    [payload removeObjectForKey:kMFModifiedDragDictKeyType];
    _addModePayload = payload;
    _didConclude = NO;
}

+ (void)handleBecameInUseWithEvent:(CGEventRef)event {
    
    return;
}

+ (void)handleMouseInputWhileInUseWithDeltaX:(double)deltaX deltaY:(double)deltaY event:(nonnull CGEventRef)event {
    
    if (!_didConclude) {
        if (_addModePayload != nil) {
            [Remap sendAddModeFeedback:_addModePayload];
            _didConclude = YES;
        } else {
            @throw [NSException exceptionWithName:@"InvalidAddModeFeedbackPayload" reason:@"_drag.addModePayload is nil. Something went wrong!" userInfo:nil]; /// Throw exception to cause crash
        }
    }
}

+ (void)handleDeactivationWhileInUseWithCancel:(BOOL)cancelation {
}

+ (void)suspend {}
+ (void)unsuspend {}

@end

File: Helper/Core/Drag/AddMode/ModifiedDragOutputAddMode.swift
//
// --------------------------------------------------------------------------
// ModifiedDragOutputAddMode.swift
// Created for Mac Mouse Fix (https://github.com/noah-nuebling/mac-mouse-fix)
// Created by Noah Nuebling in 2022 (Translated to Swift in 2026)
// Licensed under the MMF License (https://github.com/noah-nuebling/mac-mouse-fix/blob/master/License)
// --------------------------------------------------------------------------
//

import Foundation
import CoreGraphics

@objc(ModifiedDragOutputAddMode)
class ModifiedDragOutputAddMode: NSObject, ModifiedDragOutputPlugin {
    
    private weak var drag: ModifiedDragState?
    private var addModePayload: [AnyHashable: Any]?
    private var didConclude: Bool = false
    
    func initialize(with dragState: ModifiedDragState) {
        self.drag = dragState
        self.didConclude = false
        
        if let effect = drag?.effectDict {
            var payload = effect
            payload.removeValue(forKey: kMFModifiedDragDictKeyType)
            self.addModePayload = payload
        }
    }
    
    func handleBecameInUse() {}
    
    func handleMouseInputWhileInUse(deltaX: Double, deltaY: Double, event: CGEvent?) {
        if !didConclude {
            if let payload = addModePayload {
                Remap.sendAddModeFeedback(payload as NSDictionary)
                didConclude = true
            } else {
                fatalError("InvalidAddModeFeedbackPayload: _drag.addModePayload is nil.")
            }
        }
    }
    
    func handleDeactivationWhileInUse(cancel: Bool) {}
    func suspend() {}
    func unsuspend() {}
}
